### PR TITLE
groups: secret groups

### LINKED ISSUE #729: groups: implement secret groups
https://www.figma.com/file/sCFYLlAGjehlYXUkT2rarg/Product-Handoffs-(Locked)?node-id=1207%3A42843

https://www.figma.com/file/CZnMVcDGH4Ix0WWDMVfD3J/2022-Q2-Groups?node-id=6462%3A207521

### PR DESCRIPTION
**Pull request by tomholford**
# Context

This resolves #729

# Changes

## desk
- Add `secret` boolean to `group` and `preview` structs
- Add support for `secret` to update actions and JSON encoder / parser
- Update permissions model for viewing previews

## ui
- Update corresponding frontend types and state actions
- Consume `secret` boolean when rendering and interacting with groups
- Refactor `privacy` behavior
- Fix edit Group Info flow

### Find Groups (secret filtered from results)

image

#### In the group

image

#### Not in the group

image

#### Editable in Group Info view

image

# TODO
- [ ] refine [group preview viewership logic](https://github.com/tloncorp/landscape-apps/pull/969/files#diff-af9126852398aec6703fd6e760bfd8cb2460762ceb5a7356f56622737755b071R535) (i.e., if a user has been invited to a group, should be able to view the group preview)
- [ ] per discussion with @arthyn, will need state migration logic before merging
- [x] how to handle changing a group from public <--> private? (the cordon will need to be migrated, currently there is no appropriate backend action)

### CODE REVIEW

**Comment by nerveharp** [comment]
Prevent the secret group from showing up in the GangIndex.

File: ui/src/groups/GroupAdmin/GroupInfoEditor.tsx [hunk 1]
```diff
@@ -40,17 +40,18 @@ export default function GroupInfoEditor({ title }: ViewProps) {
...
   });
 
-  useEffect(() => {
-    form.reset({
-      ...emptyMeta,
-      ...group?.meta,
-      privacy: (group && getGroupPrivacy(group.cordon)) || 'public',
-    });
-  }, [group, form]);
+  // TODO: is this still needed? don't want to get Chesterton's Fence'd
```
Comment (tomholford) [inline ui/src/groups/GroupAdmin/GroupInfoEditor.tsx:47]: @arthyn Do we still need this? Don't want to get Chesterton's Fence'd :)

File: ui/src/components/References/GroupReference.tsx [hunk 2]
```diff
@@ -1,29 +1,32 @@
...
+  const { privacy } = useGroupPrivacy(flag);
+
   useEffect(() => {
     if (!gang?.preview && !group) {
       useGroupState.getState().search(flag);
     }
   }, [gang, group, flag]);
 
-  if (privacy === 'secret') {
+  if (!group && privacy === 'secret') {
```
Comment (arthyn) [inline ui/src/components/References/GroupReference.tsx:29]: is this supposed to be `&&` or should it be `||`?

File: ui/src/groups/GroupAdmin/GroupInfoEditor.tsx (hunk 1 above)
Comment (arthyn) [inline ui/src/groups/GroupAdmin/GroupInfoEditor.tsx:47]: I believe we still need this we need to reload the form with the correct privacy if it's not immediately available on first mount.

File: ui/src/groups/GroupAdmin/GroupInfoEditor.tsx [hunk 3]
```diff
@@ -67,8 +68,11 @@ export default function GroupInfoEditor({ title }: ViewProps) {
   }, [groupFlag, navigate]);
 
   const onSubmit = useCallback(
-    (values: GroupMeta) => {
-      useGroupState.getState().edit(groupFlag, values);
+    async (values: GroupMeta & { privacy: PrivacyType }) => {
+      await useGroupState.getState().edit(groupFlag, values);
```
Comment (arthyn) [inline ui/src/groups/GroupAdmin/GroupInfoEditor.tsx:72]: This doesn't full address privacy setting, it needs to also set the cordon appropriately like we do in the NewGroup form

**Review by arthyn — CHANGES_REQUESTED** [review]
A few changes, either @liam-fitzgerald or I should probably pair with you on your TODOs

File: ui/src/components/References/GroupReference.tsx (hunk 2 above)
Comment (tomholford) [reply ui/src/components/References/GroupReference.tsx]: Realized that the new hook `useGroupPrivacy` accounts for the case where there is no `group` but there is a `gang`, so this logic can be simplified.

**Comment by arthyn** [comment]
@ryjm @liam-fitzgerald pinging y'all for review

File: desk/app/groups.hoon [hunk 4]
```diff
@@ -352,7 +448,8 @@
     =-  (fall - [(crip "{(scow %p p.flag)}/{(scow %ta q.flag)}") '' '' ''])
     (bind (~(get by om) [%groups flag]) old-assoc-to-new-meta)
   =/  =group:g
-    [fleet cabals zones zone-ord bloc channels cordon meta]
+    :: TODO: rework migration for secret group
```
Comment (ryjm) [inline desk/app/groups.hoon:451]: is this needed still?

File: desk/app/groups.hoon [hunk 5]
```diff
@@ -105,7 +201,7 @@
       ^-  vessel:fleet:g
       [sects *time]
     =/  =group:g
-      [fleet ~ ~ ~ ~ ~ cordon.create title.create description.create image.create cover.create] 
+      [fleet ~ ~ ~ ~ ~ cordon.create secret.create title.create description.create image.create cover.create]
```
Comment (ryjm) [inline desk/app/groups.hoon:204]: maybe time to `:*` this bad boi

File: desk/app/groups.hoon [hunk 6]
```diff
@@ -32,14 +32,110 @@
...
+    ++  preview-0-to-1
+      |=  =preview:zero
+      ^-  preview:one
+      :*  flag.preview
+          meta.preview
+          cordon.preview
+          time.preview
+          |
+      ==
+    --
```
Comment (ryjm) [inline desk/app/groups.hoon:138]: registering my displeasure in all of this just to add a single field, but oh well. once we get past this deadline i vow to make you all `$&` maximalists.

**Review by ryjm — APPROVED** [review]
hoon side lgtg.

can feel the weight of all these state updates already.. we're going to have 5 `sur` files for each app pretty soon.

**Review by arthyn — APPROVED** [review]
front-end lgtm

**Comment by tomholford** [comment]
@ryjm Thanks for the hoon feedbck, addressed in the most recent commit

File: desk/app/groups.hoon [hunk 7]
```diff
@@ -201,7 +201,19 @@
       ^-  vessel:fleet:g
       [sects *time]
     =/  =group:g
-      [fleet ~ ~ ~ ~ ~ cordon.create secret.create title.create description.create image.create cover.create] 
+      :*  fleet
+          ~
```
Comment (ryjm) [inline desk/app/groups.hoon:205]: can put this all on one line

**Review by ryjm — CHANGES_REQUESTED** [review]
~~also see #1031 (@liam-fitzgerald made some comments on the state migration there, which i mostly copied from here)~~

EDIT: nvm no relevant comments in the state migration

File: desk/app/groups.hoon (hunk 7 above)
Comment (ryjm) [reply desk/app/groups.hoon:205]: the sigs i mean

File: desk/app/groups.hoon [hunk 8]
```diff
@@ -32,14 +32,110 @@
   ++  on-save  !>(state)
   ++  on-load
```
Comment (ryjm) [inline desk/app/groups.hoon:33]: move to `load` arm (see #1031 diary state migration

**Comment by tomholford** [comment]
@ryjm Thanks for the additional feedback, addressed in most recent commit

**Pull request merged**